lib/semantics/assignment.cpp

```diff
@@ -333,14 +218,68 @@ void AssignmentContext::CheckForPureContext(const SomeExpr &lhs,
   }
 }
 
-MaskExpr AssignmentContext::GetMask(
-    const parser::LogicalExpr &logicalExpr, bool defaultValue) {
-  MaskExpr mask{defaultValue};
-  if (const SomeExpr * expr{GetExpr(logicalExpr)}) {
-    auto *logical{std::get_if<evaluate::Expr<evaluate::SomeLogical>>(&expr->u)};
-    mask = evaluate::ConvertTo(mask, common::Clone(DEREF(logical)));
+// 10.2.3.1(2) The masks and LHS of assignments must all have the same shape
+void AssignmentContext::CheckShape(parser::CharBlock at, const SomeExpr *expr) {
+  if (auto shape{evaluate::GetShape(foldingContext(), expr)}) {
+    std::size_t size{shape->size()};
+    if (whereDepth_ == 0) {
+      whereExtents_.resize(size);
+    } else if (whereExtents_.size() != size) {
+      Say(at,
+          "Must have rank %jd to match prior mask or assignment of"
+          " WHERE construct"_err_en_US,
+          whereExtents_.size());
+      return;
+    }
+    for (std::size_t i{0}; i < size; ++i) {
+      if (std::optional<std::int64_t> extent{evaluate::ToInt64((*shape)[i])}) {
+        if (!whereExtents_[i]) {
+          whereExtents_[i] = *extent;
+        } else if (*whereExtents_[i] != *extent) {
+          Say(at,
+              "Dimension %d must have extent %jd to match prior mask or"
```

Comment:
`%jd` is for `intmax_t`, which we should not assume is the same integral type as `int64_t`.

Reply:
So in this case we have to cast to `intmax_t`, right?

Reply:
That's what I typically do, yes.  It's ugly but it should be future-proof.

Reply:
Or pass a string.  Do we have an itoa?